accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,375 @@
+
+Example:
+
+```dart
+extension BoxCom<T extends Comparable<T>> on Box<Iterable<T>> { foo() {} }
+extension BoxList<T> on Box<List<T>> { foo() {} }
+extension BoxSpec on Box<List<num>> { foo() {} }
+...
+  List<int> x = ...;
+  x.foo();
+```
+
+Here all three extensions apply. The most specific one is `BoxSpec` which is specialized for lists of numbers. If `BoxSpec` had not been in scope, then neither of `BoxCom` or `BoxList` would be more specific than the other. Their `on` types, when instantiated to bounds, are `Box<Iterable<Comparable<dynamic>>>` and `Box<List<dynamic>>` which are unrelated by subtyping.
+
+This is also why we use the instantiated-to-bounds type for comparison. If we used the actual type, bound to the type variable by the current use-case, we would consider `BoxList<int>` to be more specific than `BoxSpec`, but in practice, the `BoxList<int>` extension would not be able to *use* the integer-ness of the type in its code, so `BoxSpec`, which can specialize its code for lists of numbers, is the most precise match.
+
+### Overriding Access
+
+If two or more extensions apply to the same member access, or if a member of the receiver type prevents using an extension method, or if the extension is shadowed, then it is possible to force an extension member invocation:
+
+```dart
+MyList(object).quickSort();
+```
+
+The syntax looks like a constructor invocation, but it does not create a new object.
+
+If `object.quickSort()` would invoke an extension method of `MyList`, then `MyList(object).quickSort()` will invoke the exact same method in the same way.
+
+This mode of invocation also allows invocation of extensions that are *not* in scope, perhaps as `prefix.MyExtension(object).extensionMember()` if imported with a prefix.
+
+The syntax is not *convenient*&mdash;you have to put the "constructor" invocation up front, which removes the one advantage that extension methods have over normal static methods. It is not intended as the common use-case, but as an escape hatch out of conflicts.
+
+### Static Members and Member Resolution
+
+Static member declarations in the extension declaration can be accessed the same way as static members of a class or mixin declaration: By prefixing with the extension's name.
+
+Example:
+
+```dart
+extension MySmart on Object {
+  smart() => smartHelper(this);  // valid
+  static smartHelper(Object o) { ... }
+}
+...
+  MySmart.smartHelper(someObject);  // valid
+```
+
+### Semantics of Invocations
+
+If an extension is found to be the one applying to a member invocation, then at run-time, the invocation will perform a method invocation of the corresponding instance member of the extension, with `this` bound to the receiver value and type parameters bound to the types found by static inference.
+
+If the receiver is `null`, then that invocation is an immediate run-time error unless the `on` type of the extension has a trailing `?`.
```

Comment:
Ah, I maybe understand your previous comment better - you're not assuming NNBD.  In post NNBD world, does this mean "unless the on type is nullable"?  Or "unless the on type is potentially nullable"?

e.g.

```dart
extension E1<T extends Object?> on T {
  void m() {};
}
extension E2 on FutureOr<Object?> {
  void n() {};
}
void test() {
  null.m(); // OK?
  null.n(); // OK?
}
```

Reply:
In NNBD-world, if the `on` type is nullable, then it applies to a nullable static receiver type, if not, it doesn't. There won't be any unsafe `null` errors (except in the presence of legacy unsafe-nullable `*` types).

So, the example is OK. Since `null` is a subtype of `Object?` and of `FutureOr<Object?>`, the `n` and `m` extension methods match the `null.m()` (with `T` inferred as `Null`) and `null.n()` invocations.

Since the `on` types are nullable, the `this` references inside the methods will be nullable as well, so it's no surprise that they might be `null`.

If the receiver type is potentially nullable

So, using NNBD:
```dart
extension E1 on Object { m1() { assert(this != null); } }
extension E2 on Object?  { m2() {} }
void main() {
  Object o1 = 1;
  o1.m1(); // OK, Object <: Object
  o1.m2(); // OK, Object <: Object?
  Object? p1 = 1;
  p1.m1(); // compile-time error, there is no m1 method matching Object?.
  p1.m2(); // OK, Object? <: Object?
  Object? p2 = null;
  p2.m1(); // compile-time error, there is no m1 method matching Object?.
  p2.m2(); // OK, Object? <: Object?.
}
```

Using non-NNBD semantics, where the `?` on the type means "accepts null":
```dart
extension E1 on Object { m1() { assert(this != null); } }
extension E2 on Object?  { m2() {} }
void main() {
  Object o1 = 1;
  o1.m1(); // OK, o1 != null
  o1.m2(); // OK.
  Object? o2 = null;
  o2.m1(); // run-time error, o2 == null at runtime.
  o2.m2(); // OK, o2 == null at runtime is accepted and this == null.
}
```

Reply:
Just to be sure I understand correctly:

pre-NNBD:

```dart
extension E<T> on T {
  void foo(){}
}

void main() {
  null.foo();
}
```

is a runtime error, but post-NNBD will be accepted?

Reply:
Yes. If you *want* the pre-migration behavior, you have to migrate it to:
```dart
extension E<T extends Object> on T {
  void foo() {}
}
```
(which is also valid pre-migration and does the same thing),
and if you want the post-NNBD behavior prior to migration, you have to write;
```dart
extension E<T> on T? {
  void foo() {}
}
```
(which will still work post-migration, the `?` will just be redundant).